### PR TITLE
Symbol statistics for arithmetic and coalesce expressions"

### PR DESCRIPTION
**Pull request by kokosing**

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/cost/EstimateAssertion.java [hunk 1]
```diff
@@ -25,8 +23,10 @@ private EstimateAssertion()
     {
     }
 
-    public static void assertEstimateEquals(double actual, double expected, String messageFormat, Object... mesaggeObjects)
+    public static void assertEstimateEquals(double actual, double expected, String messageFormat, Object... messageObjects)
     {
-        assertTrue(MoreMath.nearlyEqual(actual, expected, TOLERANCE), String.format(messageFormat, mesaggeObjects));
+        if (!MoreMath.nearlyEqual(actual, expected, TOLERANCE)) {
+            throw new AssertionError(String.format(messageFormat, messageObjects) + String.format(", expected [%f], but got [%f]", expected, actual));
```
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/cost/EstimateAssertion.java]: WHy not use `assertTrue` and pass `String.format(messageFormat, messageObjects) + String.format(", expected [%f], but got [%f]", expected, actual)` ass error message in second parameter?

File: presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java [hunk 2]
```diff
@@ -35,6 +35,10 @@
 import static com.facebook.presto.sql.planner.LiteralInterpreter.evaluate;
 import static java.util.Objects.requireNonNull;
 
+/**
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java]: I would drop it.
It is outdated already. It will get worse

File: presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java [hunk 3]
```diff
@@ -93,7 +93,8 @@ protected SymbolStatsEstimate visitNullLiteral(NullLiteral node, Void context)
         protected SymbolStatsEstimate visitLiteral(Literal node, Void context)
         {
             Object value = evaluate(metadata, session.toConnectorSession(), node);
-            Type type = ExpressionAnalyzer.createConstantAnalyzer(metadata, session, ImmutableList.of()).analyze(node, Scope.create());
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java]: Not sure if that is a fix. Make this a "fixup" commit please.

File: presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java [hunk 4]
```diff
@@ -106,5 +110,72 @@ protected SymbolStatsEstimate visitLiteral(Literal node, Void context)
...
+        protected SymbolStatsEstimate visitArithmeticBinary(ArithmeticBinaryExpression node, Void context)
+        {
+            requireNonNull(node, "node is null");
+            SymbolStatsEstimate left = process(node.getLeft());
+            SymbolStatsEstimate right = process(node.getRight());
+
+            SymbolStatsEstimate.Builder result = SymbolStatsEstimate.builder()
+                    .setDataSize(Math.max(left.getDataSize(), right.getDataSize()))
+                    .setNullsFraction(left.getNullsFraction() + right.getNullsFraction() - left.getNullsFraction() * right.getNullsFraction())
+                    .setDistinctValuesCount(min(left.getDistinctValuesCount() * right.getDistinctValuesCount(), input.getOutputRowCount()));
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java]: add `TODO` to drop the `min ... input.getOutputRowCount` when we have generic rule to do that after each computation.

File: presto-main/src/test/java/com/facebook/presto/cost/TestScalarStatsCalculator.java [hunk 5]
```diff
@@ -86,4 +91,131 @@ public void testLiteral()
...
+        assertCalculate(expression("x * y"), relationStats)
+                .distinctValuesCount(10.0)
+                .lowValue(-20.0)
+                .highValue(50.0)
+                .nullsFraction(0.28)
+                .dataSize(10);
+    }
+
+    @Test
+    public void testDivideArithmeticBinaryExpression()
```
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/cost/TestScalarStatsCalculator.java:247]: add test with divisor range (0, non-zero) and (-non-zero, 0).

File: presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java [hunk 6]
```diff
@@ -106,5 +110,72 @@ protected SymbolStatsEstimate visitLiteral(Literal node, Void context)
...
+            double rightLow = right.getLowValue();
+            double rightHigh = right.getHighValue();
+            if (node.getType() == ArithmeticBinaryExpression.Type.DIVIDE && rightLow < 0 && rightHigh > 0) {
+                result.setLowValue(Double.NEGATIVE_INFINITY)
+                        .setHighValue(Double.POSITIVE_INFINITY);
+            }
+            else if (node.getType() == ArithmeticBinaryExpression.Type.MODULUS) {
+                double maxDivisor = max(abs(rightLow), abs(rightHigh));
+                if (leftHigh <= 0) {
+                    result.setLowValue(-min(maxDivisor, abs(leftLow)))
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java]: `setLowValue(max(-maxDivisor, leftLow))`

File: presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java [hunk 7]
```diff
@@ -106,5 +110,72 @@ protected SymbolStatsEstimate visitLiteral(Literal node, Void context)
...
+                            .setHighValue(0);
+                }
+                else if (leftLow >= 0) {
+                    result.setLowValue(0)
+                            .setHighValue(min(maxDivisor, leftHigh));
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java:193]: `setHighValue(min(maxDivisor, leftHigh))`

File: presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java [hunk 8]
```diff
@@ -106,5 +110,72 @@ protected SymbolStatsEstimate visitLiteral(Literal node, Void context)
...
+                } else {
+                    result.setLowValue(-min(maxDivisor, abs(leftLow)))
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java]: like above

**Review by losipiuk — COMMENTED** [review]
looks good. Some minor comments.
